pkg/graveler/committed/apply.go

```diff
@@ -14,205 +14,296 @@ type ApplyOptions struct {
 	AllowEmpty bool
 }
 
-func addIntoDiffSummary(d *graveler.DiffSummary, typ graveler.DiffType, n int) {
-	if d.Count != nil {
-		d.Count[typ] += n
-	}
-}
+// ReferenceType represents the type of the reference
 
-func incrementDiffSummary(d *graveler.DiffSummary, typ graveler.DiffType) {
-	addIntoDiffSummary(d, typ, 1)
+type applier struct {
+	ctx                   context.Context
+	logger                logging.Logger
+	writer                MetaRangeWriter
+	source                Iterator
+	diffs                 Iterator
+	opts                  *ApplyOptions
+	summary               graveler.DiffSummary
+	haveDiffs, haveSource bool
 }
 
-// ReferenceType represents the type of the reference
-
-// applyFromSource applies all changes from source to writer.
-func applyFromSource(ctx context.Context, logger logging.Logger, writer MetaRangeWriter, source Iterator) error {
+// applyAll applies all changes from Iterator to writer and returns the number of writes
+func (a *applier) applyAll(iter Iterator) (int, error) {
```

Comment:
E.g. because of the iterator/have separation above, I think this function is unsafe to call on iterator `X` unless `haveX`.